package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -408,7 +409,7 @@ def _get_dh_pairs(self):
         # If donors_sel is not provided, use topology to find d-h pairs
         if not self.donors_sel:
             if not (hasattr(self.u, 'bonds') and len(self.u.bonds) != 0):
```

Comment:
I also [posted](https://github.com/MDAnalysis/mdanalysis/issues/2396#issuecomment-596251787) this in the related issue for reference:

`hasattr(u, 'bonds')` is an *incredibly slow* test because it involves a lot of object instantiations under the hood. `hasattr(u._topology, 'bonds')` is **much** faster:
```python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import GRO, TPR

u = mda.Universe(GRO)  # universe *without* bonds
%timeit hasattr(u, 'bonds')
# 758 µs ± 4.33 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit hasattr(u._topology, 'bonds')
# 216 ns ± 2.88 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

u = mda.Universe(TPR)  # universe *with* bonds
%timeit hasattr(u, 'bonds')
# 88.8 ms ± 524 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)   <== AAARGH!
%timeit hasattr(u._topology, 'bonds')
# 86.4 ns ± 0.266 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
You see that in the latter example, `hasattr(u, 'bonds')` is **a million times slower** than `hasattr(u._topology, 'bonds')`. This may be irrelevant for a setup routine that is run only once, but when being called in a loop, `hasattr(u, 'bonds')` is a nice way of very thoroughly killing performance. :grin:

Reply:
Hi @zemanj , 

Thanks for having a look at it. You've clearly done the work here. I can see that as you say, it is significantly slower. However, in a situation like this, I would advice to prioritise the simplicity of the code over performance. I believe you point this out as well. Thanks

Reply:
Thanks for pointing this out @zemanj - that's a major difference! If the atom selection is set to update, and the topology is to be used to find donor-hydrogen pairs, then this method (and thus the check for bonds) will be called for each frame of the analysis. So it's probably worth making this change.